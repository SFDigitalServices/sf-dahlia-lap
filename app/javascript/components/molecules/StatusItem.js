import React from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'

import StatusPill from '../atoms/StatusPill'

import StatusItemShape from '../../utils/shapes/StatusItemShape'
import { getSubStatusLabel } from '../../utils/statusUtils'

const StatusDate = ({ timestamp }) => (
  <div className='status-item-date'>
    {moment.unix(timestamp).format('MMM D, YYYY')}
  </div>
)

StatusDate.propTypes = { timestamp: PropTypes.number }

const StatusItem = ({ statusItem }) => {
  const { status, substatus, comment, timestamp } = statusItem

  const substatusLabel = getSubStatusLabel(status, substatus)

  return (
    <div className='status-item'>
      <div className='status-item-header'>
        <StatusPill status={status} />
        <StatusDate timestamp={timestamp} />
      </div>
      {substatusLabel && (
        <div className='item-substatus-text'>
          {substatusLabel}
        </div>
      )}
      <div className='status-item-text'>
        {comment}
      </div>
    </div>
  )
}

StatusItem.propTypes = {
  statusItem: PropTypes.shape(StatusItemShape).isRequired
}

export default StatusItem
