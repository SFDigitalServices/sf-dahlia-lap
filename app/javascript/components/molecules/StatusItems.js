import React from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'

import StatusPill from '../atoms/StatusPill'

import StatusItemShape from '../../utils/shapes/StatusItemShape'
import { getSubStatusLabel } from '../../utils/statusUtils'

const StatusDate = ({ timeStamp }) => (
  <div className='status-item-date'>
    {moment.unix(timeStamp).format('MMMM Do YYYY')}
  </div>
)

StatusDate.propTypes = { timeStamp: PropTypes.number }

const StatusItem = ({ statusItem }) => {
  const { status, subStatus, comment, timeStamp } = statusItem

  const subStatusLabel = getSubStatusLabel(status, subStatus)

  return (
    <div className='status-items'>
      <div className='status-item'>
        <div className='status-item-header'>
          <StatusPill status={status} />
          <StatusDate timeStamp={timeStamp} />
        </div>
        {subStatusLabel && (
          <div className='item-substatus-text'>
            {subStatusLabel}
          </div>
        )}
        <div className='status-item-text'>
          {comment}
        </div>
      </div>
    </div>
  )
}

StatusItem.propTypes = { statusItem: PropTypes.shape(StatusItemShape).isRequired }

/**
 * Displays a list of status updates for the Lease-up sidebar.
 *
 * Only displays up to {limit} updates
 *
 * Note that the statusItems prop won't be sorted by this component,
 * the items should be sorted before they're passed in for performance reasons.
 */
const StatusItems = ({ statusItems, limit }) => {
  const limitedItems = statusItems.slice(0, limit)

  return (
    <div className='status-items'>
      {limitedItems.map((item, idx) => (
        <StatusItem
          key={idx}
          statusItem={item}
        />
      ))}
    </div>
  )
}

StatusItems.propTypes = {
  limit: PropTypes.number,
  statusItems: PropTypes.arrayOf(PropTypes.shape(StatusItemShape))
}

StatusItems.defaultProps = {
  limit: Number.MAX_VALUE,
  statusItems: []
}

export default StatusItems
