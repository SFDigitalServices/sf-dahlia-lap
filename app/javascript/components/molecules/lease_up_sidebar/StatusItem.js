import React from 'react'

import moment from 'moment'
import PropTypes from 'prop-types'

import StatusItemShape from '../../../utils/shapes/StatusItemShape'
import { getSubStatusLabel } from '../../../utils/statusUtils'
import StatusPill from '../../atoms/StatusPill'

const StatusDate = ({ timestamp }) => (
  <div className='status-item-date'>{moment.unix(timestamp).format('MMM D, YYYY')}</div>
)
const StatusCreatedBy = ({ createdBy }) => <div className='status-item-created-by'>{createdBy}</div>

StatusDate.propTypes = { timestamp: PropTypes.number }

const StatusItem = ({ statusItem }) => {
  const { status, substatus, comment, timestamp, created_by } = statusItem

  const substatusLabel = getSubStatusLabel(status, substatus)

  return (
    <div className='status-item'>
      <div className='status-item-header'>
        <StatusPill status={status} />
      </div>
      {substatusLabel && <div className='item-substatus-text'>{substatusLabel}</div>}
      <div className='status-item-text'>{comment}</div>
      <div className='status-item-footer'>
        <StatusCreatedBy createdBy={created_by} />
        <StatusDate timestamp={timestamp} />
      </div>
    </div>
  )
}

StatusItem.propTypes = {
  statusItem: PropTypes.shape(StatusItemShape).isRequired
}

export default StatusItem
