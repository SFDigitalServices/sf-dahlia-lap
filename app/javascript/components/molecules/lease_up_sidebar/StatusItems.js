import React from 'react'
import PropTypes from 'prop-types'

import StatusItem from './StatusItem'

import StatusItemShape from '../../../utils/shapes/StatusItemShape'

/**
 * Displays a list of status updates for the Lease-up sidebar.
 *
 * Only displays up to {limit} updates
 *
 * Note that the statusItems prop won't be sorted by this component,
 * the items should be sorted before they're passed in for performance reasons.
 */
const StatusItems = ({ statusItems, limit }) => {
  const limitedItems = statusItems.slice(0, limit)

  return (
    <div className='status-items'>
      {limitedItems.map((item, idx) => (
        <StatusItem key={idx} statusItem={item} />
      ))}
    </div>
  )
}

StatusItems.propTypes = {
  limit: PropTypes.number,
  statusItems: PropTypes.arrayOf(PropTypes.shape(StatusItemShape))
}

StatusItems.defaultProps = {
  limit: Number.MAX_VALUE,
  statusItems: []
}

export default StatusItems
