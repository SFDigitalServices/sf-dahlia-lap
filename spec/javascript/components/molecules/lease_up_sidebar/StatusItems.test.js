import React from 'react'

import { shallow } from 'enzyme'

import StatusItem from 'components/molecules/lease_up_sidebar/StatusItem'
import StatusItems from 'components/molecules/lease_up_sidebar/StatusItems'

import { mockManyStatusItems, mockStatusItem, mockStatusItems } from '../../../mocks/statusItemMock'

const getWrapper = (items, limit = undefined, height = null) =>
  shallow(<StatusItems statusItems={items} limit={limit} height={height} />)

const TIMESTAMP_AUGUST_25_2020 = 1598338800
const TIMESTAMP_AUGUST_26_2020 = 1598400000

describe('StatusItems', () => {
  describe('snapshot tests', () => {
    test('should render empty status items correctly', () => {
      const wrapper = getWrapper([])
      expect(wrapper).toMatchSnapshot()
    })

    test('should render a single status item correctly', () => {
      const wrapper = getWrapper([mockStatusItem()])
      expect(wrapper).toMatchSnapshot()
    })

    test('should render multiple status items correctly', () => {
      const wrapper = getWrapper(mockStatusItems())
      expect(wrapper).toMatchSnapshot()
    })

    test('should render multiple status items with limit 0 correctly', () => {
      const wrapper = getWrapper(mockStatusItems(), 0)
      expect(wrapper).toMatchSnapshot()
    })

    test('should render multiple status items with limit 1 correctly', () => {
      const wrapper = getWrapper(mockStatusItems(), 1)
      expect(wrapper).toMatchSnapshot()
    })

    test('should render multiple status items with limit 100 correctly', () => {
      const wrapper = getWrapper(mockStatusItems(), 100)
      expect(wrapper).toMatchSnapshot()
    })
  })

  it('should not set a height if height is not provided', () => {
    const wrapper = getWrapper(mockStatusItems(), 4)
    const styles = wrapper.find('.status-items').props().style
    expect(styles).toEqual({})
  })

  it('should limit the height of the list if provided', () => {
    const wrapper = getWrapper(mockStatusItems(), 4, '20rem')
    const styles = wrapper.find('.status-items').props().style
    expect(styles).toEqual({ height: '20rem', overflow: 'scroll' })
  })

  describe('with an empty status items list', () => {
    const statusItems = []

    test('should render no status items when no limit is specified', () => {
      expect(getWrapper(statusItems).find(StatusItem)).toHaveLength(0)
    })

    test('should render no status items when limit is 0', () => {
      expect(getWrapper(statusItems, 0).find(StatusItem)).toHaveLength(0)
    })

    test('should render no status items when limit is 1', () => {
      expect(getWrapper(statusItems, 1).find(StatusItem)).toHaveLength(0)
    })

    test('should render no status items when limit is 4', () => {
      expect(getWrapper(statusItems, 4).find(StatusItem)).toHaveLength(0)
    })
  })

  describe('with a single item in the status items list', () => {
    const statusItems = [mockStatusItem()]

    test('should render one status item when no limit is specified', () => {
      expect(getWrapper(statusItems).find(StatusItem)).toHaveLength(1)
    })

    test('should render no status items when limit is 0', () => {
      expect(getWrapper(statusItems, 0).find(StatusItem)).toHaveLength(0)
    })

    test('should render one status item when limit is 1', () => {
      expect(getWrapper(statusItems, 1).find(StatusItem)).toHaveLength(1)
    })

    test('should render one status item when limit is 4', () => {
      expect(getWrapper(statusItems, 4).find(StatusItem)).toHaveLength(1)
    })
  })

  describe('with five items in the status items list', () => {
    const statusItems = mockManyStatusItems(5)

    test('should render five status items when no limit is specified', () => {
      expect(getWrapper(statusItems).find(StatusItem)).toHaveLength(5)
    })

    test('should render no status items when limit is 0', () => {
      expect(getWrapper(statusItems, 0).find(StatusItem)).toHaveLength(0)
    })

    test('should render one status items when limit is 1', () => {
      expect(getWrapper(statusItems, 1).find(StatusItem)).toHaveLength(1)
    })

    test('should render four status items when limit is 4', () => {
      expect(getWrapper(statusItems, 4).find(StatusItem)).toHaveLength(4)
    })
  })

  describe('with two status items in chronological order', () => {
    const statusItems = [
      mockStatusItem({ timestamp: TIMESTAMP_AUGUST_25_2020 }),
      mockStatusItem({ timestamp: TIMESTAMP_AUGUST_26_2020 })
    ]

    test('should render status items in the order specified', () => {
      const statusItemsWrapper = getWrapper(statusItems).find(StatusItem)
      const firstStatusItem = statusItemsWrapper.first()
      const secondStatusItem = statusItemsWrapper.at(1)
      expect(firstStatusItem.prop('statusItem').timestamp).toEqual(TIMESTAMP_AUGUST_25_2020)
      expect(secondStatusItem.prop('statusItem').timestamp).toEqual(TIMESTAMP_AUGUST_26_2020)
    })
  })

  describe('with two status items not in chronological order', () => {
    const statusItems = [
      mockStatusItem({ timestamp: TIMESTAMP_AUGUST_26_2020 }),
      mockStatusItem({ timestamp: TIMESTAMP_AUGUST_25_2020 })
    ]

    test('should render status items in the order specified', () => {
      const statusItemsWrapper = getWrapper(statusItems).find(StatusItem)
      const firstStatusItem = statusItemsWrapper.first()
      const secondStatusItem = statusItemsWrapper.at(1)
      expect(firstStatusItem.prop('statusItem').timestamp).toEqual(TIMESTAMP_AUGUST_26_2020)
      expect(secondStatusItem.prop('statusItem').timestamp).toEqual(TIMESTAMP_AUGUST_25_2020)
    })
  })
})
