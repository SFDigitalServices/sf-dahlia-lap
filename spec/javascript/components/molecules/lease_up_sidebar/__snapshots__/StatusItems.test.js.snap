// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusItems snapshot tests should render a single status item correctly 1`] = `
<div
  className="status-items"
>
  <StatusItem
    key="0"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Approved",
        "substatus": "Approval letter sent",
        "timestamp": 1598385600,
      }
    }
  />
</div>
`;

exports[`StatusItems snapshot tests should render empty status items correctly 1`] = `
<div
  className="status-items"
/>
`;

exports[`StatusItems snapshot tests should render multiple status items correctly 1`] = `
<div
  className="status-items"
>
  <StatusItem
    key="0"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Approved",
        "substatus": "Approval letter sent",
        "timestamp": 1598385600,
      }
    }
  />
  <StatusItem
    key="1"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Approved",
        "substatus": "Unit selected",
        "timestamp": 1598299200,
      }
    }
  />
  <StatusItem
    key="2"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Approved",
        "substatus": "Waiting for subsidy inspection",
        "timestamp": 1598212800,
      }
    }
  />
  <StatusItem
    key="3"
    statusItem={
      Object {
        "comment": null,
        "status": "Waitlisted",
        "substatus": "Written confirmation sent",
        "timestamp": 1598126400,
      }
    }
  />
  <StatusItem
    key="4"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Lease Signed",
        "substatus": null,
        "timestamp": 1598040000,
      }
    }
  />
  <StatusItem
    key="5"
    statusItem={
      Object {
        "comment": null,
        "status": "Lease Signed",
        "substatus": null,
        "timestamp": 1597953600,
      }
    }
  />
</div>
`;

exports[`StatusItems snapshot tests should render multiple status items with limit 0 correctly 1`] = `
<div
  className="status-items"
/>
`;

exports[`StatusItems snapshot tests should render multiple status items with limit 1 correctly 1`] = `
<div
  className="status-items"
>
  <StatusItem
    key="0"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Approved",
        "substatus": "Approval letter sent",
        "timestamp": 1598385600,
      }
    }
  />
</div>
`;

exports[`StatusItems snapshot tests should render multiple status items with limit 100 correctly 1`] = `
<div
  className="status-items"
>
  <StatusItem
    key="0"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Approved",
        "substatus": "Approval letter sent",
        "timestamp": 1598385600,
      }
    }
  />
  <StatusItem
    key="1"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Approved",
        "substatus": "Unit selected",
        "timestamp": 1598299200,
      }
    }
  />
  <StatusItem
    key="2"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Approved",
        "substatus": "Waiting for subsidy inspection",
        "timestamp": 1598212800,
      }
    }
  />
  <StatusItem
    key="3"
    statusItem={
      Object {
        "comment": null,
        "status": "Waitlisted",
        "substatus": "Written confirmation sent",
        "timestamp": 1598126400,
      }
    }
  />
  <StatusItem
    key="4"
    statusItem={
      Object {
        "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
        "status": "Lease Signed",
        "substatus": null,
        "timestamp": 1598040000,
      }
    }
  />
  <StatusItem
    key="5"
    statusItem={
      Object {
        "comment": null,
        "status": "Lease Signed",
        "substatus": null,
        "timestamp": 1597953600,
      }
    }
  />
</div>
`;
